pulsar-client/src/main/java/zipkin2/reporter/pulsar/PulsarSender.java

```diff
@@ -0,0 +1,369 @@
+    return encoding.listSizeInBytes(encodedSpans);
+  }
+
+  @Override public int messageSizeInBytes(int encodedSizeInBytes) {
+    return encoding.listSizeInBytes(encodedSizeInBytes);
+  }
+
+  @Override public Encoding encoding() {
+    return encoding;
+  }
+
+  @Override public int messageMaxBytes() {
+    return messageMaxBytes;
+  }
+
+  /** {@inheritDoc} */
+  @Override @Deprecated public Call<Void> sendSpans(List<byte[]> encodedSpans) {
+    if (closeCalled) throw new ClosedSenderException();
+    byte[] message = encoding.encode(encodedSpans);
+    return new PulsarCall(message);
+  }
+
+  @Override public void send(List<byte[]> encodedSpans) {
+    if (closeCalled) throw new ClosedSenderException();
+    sender(encoding.encode(encodedSpans));
+  }
+
+  void sender(byte[] message) {
+    if (closeCalled) throw new ClosedSenderException();
+    sendMessage(message);
+  }
+
+  void sendMessage(byte[] message) {
+    if (client == null) {
+      synchronized (this) {
+        if (client == null) {
+          client = createClient();
+          producer = createProducer(client);
+
+          try {
+            producer.newMessage()
+              .value(message)
+              .loadConf(messageProps)
+              .sendAsync();
+          } catch (Exception e) {
+            cleanup();
+            throw new RuntimeException("Pulsar producer send failed." + e.getMessage(), e);
+          }
+        }
+      }
+    }
+  }
+
+  Producer<byte[]> createProducer(PulsarClient client) {
+    try {
+      producer = client.newProducer()
+        .topic(topic)
+        .enableBatching(false) // disabling batching as duplicates effort covered by sender buffering.
+        .loadConf(producerProps)
+        .create();
+    } catch (Exception e) {
+      cleanup();
+      throw new RuntimeException("Pulsar producer creation failed." + e.getMessage(), e);
+    }
+    return producer;
+  }
+
+  PulsarClient createClient() {
+    try {
+      client = PulsarClient.builder()
+        .loadConf(clientProps)
+        .build();
+    } catch (PulsarClientException e) {
+      throw new RuntimeException("Pulsar client creation failed. " + e.getMessage(), e);
+    }
+    return client;
+  }
+
+  void cleanup() {
+    try {
+      if (producer != null) {
+        producer.close();
+        producer = null;
+      }
+    } catch (PulsarClientException ignored) {
+    }
+    try {
+      if (client != null) {
+        client.close();
+        client = null;
+      }
+    } catch (PulsarClientException ignored) {
+    }
+  }
+
+  /** {@inheritDoc} */
+  @Override @Deprecated public CheckResult check() {
+    try {
+      client.getPartitionsForTopic(topic).get(15, TimeUnit.SECONDS);
+    } catch (Throwable t) {
+      Call.propagateIfFatal(t);
+      return CheckResult.failed(t);
+    }
+    return CheckResult.OK;
+  }
+
+  @Override public synchronized void close() throws IOException {
+    if (closeCalled) return;
+    PulsarClient client = this.client;
```

Comment:
I think you could now reuse cleanup here:

```suggestion
    cleanup();
```

Reply:
oops, I miss it, let me adjust this.